<a href="https://colab.research.google.com/github/zookeeper464/data-analysis/blob/main/section2/project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **주제 부동산 시장에 대한 데이터 분석**
**분석 내용 - 수도권 부동산 투자 가치 조사**
1. 2012~2020까지의 데이터를 통해 2018의 테스트 데이터를 조사한다.(3456개)
  2020 서울지역의 (아파트,연립,단독)을 조사한다.(O)
2. 중위 가격비, 매매가격을 통해 실투자금에 대해 조사한다.(O)
3. 2년동안의 중위 가격비와 주택 담보대출 금리를 비교하여 계산값을 도출한다.(O)
4. 2년 동안 금리와 비교하여 3.의 column이 더 큰지 작은지 비교한다.(O)
5. 이를 통해 타겟을 설정하고 타겟은 투자여부로 설정한다.(O)
---
**피쳐설정**
1. 실거래지수를 통해 매매와 전세 사이의 관계 설정(O)
2. 가계대출 대비 주택 대출의 비율의 수치 설정(O)
3. 환율, 종합주가 지수, 주택 담보대출 금리, 정기 예금 금리(O)
---
**추가해야 하는 내용**

선택한 모델에 대한 이유 설명
1. 왜 해당 데이터를 선택했는지 자세히 설명한다.(O)
2. 가설은 쓸모 있어야 한다.(O)
3. 해당 특성을 target으로 지정한 이유를 설명하세요.(O)
4. 해당 베이스라인 모델과 평가지표를 선택한 이유를 설명하세요.(O)
5. 만들고자 하는 모델의 한계와 모델의 목표 설명하세요.(O)
6. Data Leakage가 있었나요? 없었다면 어떻게 방지했나요?(O)
7. 김하진님이 만들고자 하는 모델은 언제 유용한가요? 어떤 한계를 가지고 있나요?(O)

모델 학습과 관련한 내용
8. 다양한 모델을 학습하세요. (O)
9. 평가지표를 계산 후 베이스라인과 비교해보세요.(O)
10. 어느정도 성능이 나왔다면, 교차 검증 (이하 CV)을 통해서 일반화될 가능성이 있는지 확인해봅니다.(O)
11. 최소 2개 이상의 모델을 만들어서 validation 점수를 보고하세요.(O)
12. 최종 모델의 test 점수를 보고하세요.(O)

모델 검증에 대한 내용
13. 모델을 학습한 후에 베이스라인보다 잘 나왔나요? 그렇지 않다면 그 이유는 무엇일까요?(O)
14. 모델 성능 개선을 위해 어떤 방법을 적용했나요? 그 방법을 선택한 이유는 무엇인가요?(O)
15. 최종 모델에 관해 설명하세요.()

발표에 대한 내용
16. PDP, SHAP을 활용하여 최종 모델을 설명합니다.()
17. 시각화는 "설명"이 제일 중요합니다.()
18. 모델이 관측치를 예측하기 위해서 어떤 특성을 활용했나요?()
19. 어떤 특성이 있다면 모델의 예측에 도움이 될까요? 해당 특성은 어떻게 구할 수 있을까요?()
---
**발표내용**

지금까지 학습한 내용을 토대로 주택 투자에 대한 이익에 대해서 설명하고 앞으로 변화할 주택 시장에 대해서 발표한다.()

## 기존의 데이터를 가지고 온 뒤 새롭게 전처리 하는 과정을 거친다.

In [3]:
import pandas as pd


매매대비전세_단독 = pd.read_excel("/content/매매가격대비전세가격_중위가격_단독.xlsx")
매매대비전세_아파트 = pd.read_excel("/content/매매가격대비전세가격_중위가격_아파트.xlsx")
매매대비전세_연립다세대 = pd.read_excel("/content/매매가격대비전세가격_중위가격_연립다세대.xlsx")
매매대비전세_종합 = pd.read_excel("/content/매매가격대비전세가격_중위가격_종합.xlsx")

매매가격지수_단독 = pd.read_excel("/content/월간_매매가격지수_단독.xlsx")
매매가격지수_아파트 = pd.read_excel("/content/월간_매매가격지수_아파트.xlsx")
매매가격지수_연립다세대 = pd.read_excel("/content/월간_매매가격지수_연립다세대.xlsx")
매매가격지수_종합 = pd.read_excel("/content/월간_매매가격지수_종합.xlsx")
#매매가격지수는 변동율을 제거하는 전처리를 해야한다.

정기예금금리 = pd.read_excel("/content/정기예금금리.xlsx")
종합주가지수 = pd.read_excel("/content/종합주가지수.xlsx")
주택담보대출금리 = pd.read_excel("/content/주택담보대출금리.xlsx")
주택대출 = pd.read_excel("/content/주택대출.xlsx")
가계대출 = pd.read_excel("/content/가계대출.xlsx")
환율 = pd.read_excel("/content/환율.xlsx")

In [4]:
def organized (df):
  df_t = df.transpose()
  return df_t[df_t[0] != "변동률"]

매매가격지수_단독_t = organized(매매가격지수_단독).reset_index()
매매가격지수_아파트_t = organized(매매가격지수_아파트).reset_index()
매매가격지수_연립다세대_t = organized(매매가격지수_연립다세대).reset_index()
매매가격지수_종합_t = organized(매매가격지수_종합).reset_index()
# 매매가격지수에 들어가 있는 변동률 제거 및 트랜스포즈, reset_index

매매가격지수_단독_t.columns=["년월", "val","서울",	'강북지역',	'도심권',	'동북권',	'서북권',	'강남지역',	'서남권',	"동남권"]
매매가격지수_단독_n = 매매가격지수_단독_t.drop(매매가격지수_단독_t.index[0]).drop("val", axis=1)
매매가격지수_아파트_t.columns=["년월", "val","서울",	'강북지역',	'도심권',	'동북권',	'서북권',	'강남지역',	'서남권',	"동남권"]
매매가격지수_아파트_n = 매매가격지수_아파트_t.drop(매매가격지수_아파트_t.index[0]).drop("val", axis=1)
매매가격지수_연립다세대_t.columns=["년월", "val","서울",	'강북지역',	'도심권',	'동북권',	'서북권',	'강남지역',	'서남권',	"동남권"]
매매가격지수_연립다세대_n = 매매가격지수_연립다세대_t.drop(매매가격지수_연립다세대_t.index[0]).drop("val", axis=1)
매매가격지수_종합_t.columns=["년월", "val","서울",	'강북지역',	'도심권',	'동북권',	'서북권',	'강남지역',	'서남권',	"동남권"]
매매가격지수_종합_n = 매매가격지수_종합_t.drop(매매가격지수_종합_t.index[0]).drop("val", axis=1)
#columns와 index를 사용하기 좋게 통일시켰다.

매매가격지수_단독_tidy = pd.melt(매매가격지수_단독_n,["년월"],var_name="지역",value_name="매매가격지수")
매매가격지수_아파트_tidy = pd.melt(매매가격지수_아파트_n,["년월"],var_name="지역",value_name="매매가격지수")
매매가격지수_연립다세대_tidy = pd.melt(매매가격지수_연립다세대_n,["년월"],var_name="지역",value_name="매매가격지수")
매매가격지수_종합_tidy = pd.melt(매매가격지수_종합_n,["년월"],var_name="지역",value_name="매매가격지수")

매매가격지수_단독_tidy["종류"] = "단독"
매매가격지수_아파트_tidy["종류"] = "아파트"
매매가격지수_연립다세대_tidy["종류"] = "연립다세대"
매매가격지수_종합_tidy["종류"] = "종합"
#통일시킨 데이터를 가지고 tidy형태로 변형시켰다.

In [5]:
매매대비전세_단독_t =매매대비전세_단독.T.reset_index()
매매대비전세_아파트_t = 매매대비전세_아파트.T.reset_index()
매매대비전세_연립다세대_t = 매매대비전세_연립다세대.T.reset_index()
매매대비전세_종합_t = 매매대비전세_종합.T.reset_index()

정기예금금리_t = 정기예금금리.T.reset_index()
종합주가지수_t = 종합주가지수.T.reset_index()
주택담보대출금리_t = 주택담보대출금리.T.reset_index()
주택대출_t = 주택대출.T.reset_index()
가계대출_t = 가계대출.T.reset_index()
환율_t = 환율.T.reset_index()
#나머지 특성들에게 트랜스포즈 및 reset_index

매매대비전세_단독_t.columns = ["년월","서울",	'강북지역',	'도심권',	'동북권',	'서북권',	'강남지역',	'서남권',	"동남권"]
매매대비전세_아파트_t.columns = ["년월","서울",	'강북지역',	'도심권',	'동북권',	'서북권',	'강남지역',	'서남권',	"동남권"]
매매대비전세_연립다세대_t.columns = ["년월","서울",	'강북지역',	'도심권',	'동북권',	'서북권',	'강남지역',	'서남권',	"동남권"]
매매대비전세_종합_t.columns = ["년월","서울",	'강북지역',	'도심권',	'동북권',	'서북권',	'강남지역',	'서남권',	"동남권"]
매매대비전세_단독_t = 매매대비전세_단독_t.drop(매매대비전세_단독_t.index[0])
매매대비전세_아파트_t = 매매대비전세_아파트_t.drop(매매대비전세_아파트_t.index[0])
매매대비전세_연립다세대_t = 매매대비전세_연립다세대_t.drop(매매대비전세_연립다세대_t.index[0])
매매대비전세_종합_t = 매매대비전세_종합_t.drop(매매대비전세_종합_t.index[0])

정기예금금리_t.columns = ["년월","정기예금금리"]
종합주가지수_t.columns = ["년월","종합주가지수"]
주택담보대출금리_t.columns = ["년월", "주택담보대출금리"]
주택대출_t.columns = ["년월", "주택대출"]
가계대출_t.columns = ["년월", '가계대출']
환율_t.columns = ["년월", "환율"]
정기예금금리_t = 정기예금금리_t.drop(정기예금금리_t.index[0])
종합주가지수_t = 종합주가지수_t.drop(종합주가지수_t.index[0])
주택담보대출금리_t = 주택담보대출금리_t.drop(주택담보대출금리_t.index[0])
주택대출_t = 주택대출_t.drop(주택대출_t.index[0])
가계대출_t = 가계대출_t.drop(가계대출_t.index[0])
환율_t = 환율_t.drop(환율_t.index[0])
#columns와 index를 사용하기 좋게 통일시켰다.

매매대비전세_단독_tidy = pd.melt(매매대비전세_단독_t,["년월"],var_name="지역",value_name="매매대비전세")
매매대비전세_아파트_tidy = pd.melt(매매대비전세_아파트_t,["년월"],var_name="지역",value_name="매매대비전세")
매매대비전세_연립다세대_tidy = pd.melt(매매대비전세_연립다세대_t,["년월"],var_name="지역",value_name="매매대비전세")
매매대비전세_종합_tidy = pd.melt(매매대비전세_종합_t,["년월"],var_name="지역",value_name="매매대비전세")
매매대비전세_단독_tidy["종류"] = "단독"
매매대비전세_아파트_tidy["종류"] = "아파트"
매매대비전세_연립다세대_tidy["종류"] = "연립다세대"
매매대비전세_종합_tidy["종류"] = "종합"
#통일시킨 데이터를 가지고 tidy형태로 변형시켰다.

In [6]:
매매대비전세_종합_tidy

년월   지역   매매대비전세  종류
0    2012년 01월   서울  52.9199  종합
1    2012년 02월   서울  53.0407  종합
2    2012년 03월   서울  53.1389  종합
3    2012년 04월   서울   53.198  종합
4    2012년 05월   서울  53.2595  종합
..         ...  ...      ...  ..
859  2020년 08월  동남권   55.772  종합
860  2020년 09월  동남권  55.9273  종합
861  2020년 10월  동남권  56.2656  종합
862  2020년 11월  동남권  56.6645  종합
863  2020년 12월  동남권  57.0346  종합

[864 rows x 4 columns]

In [7]:
df1 = pd.concat([매매가격지수_종합_tidy,매매가격지수_아파트_tidy,매매가격지수_연립다세대_tidy,매매가격지수_단독_tidy]).sort_values(by=['년월',"종류"], axis=0)
df1 = df1[['년월', '지역', '종류', '매매가격지수']]
df1 = df1.reset_index().drop("index",axis=1)

df2 = pd.concat([매매대비전세_종합_tidy,매매대비전세_아파트_tidy,매매대비전세_연립다세대_tidy,매매대비전세_단독_tidy]).sort_values(by=['년월',"종류"], axis=0)
df2 = df2[['년월', '지역', '종류', '매매대비전세']]
df2 = df2.reset_index().drop("index",axis=1)
# 종류에 따른 분류가 필요한 데이터들을 모두 concat으로 처리했다.

df3 = pd.merge(df1,df2)
df3


년월    지역  종류   매매가격지수   매매대비전세
0     2012년 01월    서울  단독   96.454  41.6014
1     2012년 01월  강북지역  단독   96.404  41.8163
2     2012년 01월   도심권  단독    94.96  37.6289
3     2012년 01월   동북권  단독   96.554  41.5802
4     2012년 01월   서북권  단독   96.967   45.082
...         ...   ...  ..      ...      ...
3451  2020년 12월   동북권  종합  110.336  62.9605
3452  2020년 12월   서북권  종합  111.437   62.991
3453  2020년 12월  강남지역  종합  111.273  61.4538
3454  2020년 12월   서남권  종합  110.619  64.7898
3455  2020년 12월   동남권  종합  112.158  57.0346

[3456 rows x 5 columns]

In [8]:
#정기예금금리_t, 종합주가지수_t, 주택담보대출금리_t, 주택대출_t, 가계대출_t, 환율_t
df4 = pd.merge(정기예금금리_t,종합주가지수_t,)
df5 = pd.merge(주택담보대출금리_t,환율_t)
df5["가계대출대비주택대출"] = 주택대출_t["주택대출"]/가계대출_t["가계대출"]
df6 = pd.merge(df4,df5)
df7 = pd.merge(df6,환율_t)
df7.iloc[0,5] = 128780/200857
df8 = pd.merge(df3,df7)
df8["년월"] = pd.to_datetime(df8['년월'], format="%Y년 %m월")
df8
#데이터를 모두 취합

년월    지역  종류   매매가격지수  ...   종합주가지수 주택담보대출금리    환율 가계대출대비주택대출
0    2012-01-01    서울  단독   96.454  ...  1891.01     5.06  1125   0.641153
1    2012-01-01  강북지역  단독   96.404  ...  1891.01     5.06  1125   0.641153
2    2012-01-01   도심권  단독    94.96  ...  1891.01     5.06  1125   0.641153
3    2012-01-01   동북권  단독   96.554  ...  1891.01     5.06  1125   0.641153
4    2012-01-01   서북권  단독   96.967  ...  1891.01     5.06  1125   0.641153
...         ...   ...  ..      ...  ...      ...      ...   ...        ...
3451 2020-12-01   동북권  종합  110.336  ...  2755.77     2.59  1088   0.588427
3452 2020-12-01   서북권  종합  111.437  ...  2755.77     2.59  1088   0.588427
3453 2020-12-01  강남지역  종합  111.273  ...  2755.77     2.59  1088   0.588427
3454 2020-12-01   서남권  종합  110.619  ...  2755.77     2.59  1088   0.588427
3455 2020-12-01   동남권  종합  112.158  ...  2755.77     2.59  1088   0.588427

[3456 rows x 10 columns]

# 데이터를 저장, 필요한 데이터 처리하고, 새로운 파일을 받아 train과 test 데이터로 구분

In [9]:
#df8.to_excel(excel_writer='raw_proj_df.xlsx')

In [10]:
df = pd.read_excel("/content/proj_df.xlsx")
sub_df = df.drop(["Unnamed: 0",'2년 뒤 매매가격지수', '2년간 금리', '2년간 금리.1','판단여부'],axis=1)
df = df.drop(["Unnamed: 0",'2년 뒤 매매가격지수', '2년간 금리', '2년간 금리.1','손익분기점'],axis=1)

In [11]:
con1 = (df['년월'] < '2018-01-01')
con2 = (df['년월'] >= '2019-01-01')
con3 = ((df['년월'] < '2019-01-01') & (df['년월'] >= '2018-01-01'))

train = df.loc[con1]
test = df.loc[con2]
val = df.loc[con3]

target = '판단여부'
features = train.columns.drop([target])

X_train = train[features]
y_train = train[target]
X_val = val[features]
y_val = val[target]
X_test = test[features]

X_train_encoded = pd.get_dummies(X_train).drop("년월",axis=1)
X_test_encoded = pd.get_dummies(X_test).drop("년월",axis=1)
X_val_encoded = pd.get_dummies(X_val).drop("년월",axis=1)

#발표를 위한 주석
**해당 데이터를 선택한 이유** : 대략적인 부동산 투자 흐름에 대해서 파악하고 향후 부동산 시장의 흐름이 어떻게 바뀔지 예측하기 위해서 해당 데이터를 사용했습니다. 이 데이터를 가지고 서울의 어느 지역, 어느 지역, 어느 종류의 매물이 투자가지가 더 높은지 파악할 수 있다고 생각합니다.

**target 특성을 선택한 이유** : target특성은 제가 만든 데이터로 해당 데이터는 2년 뒤의 매물의 가격과 현재 매물 가격의 차이와 해당 기간의 대출 금리를 통해 대출한 돈으로 매물을 구매하고 전세를 놓았을 때, 추가로 지출되는 이자와 지가의 변동을 파악하고 이를 해당 기간의 금리와 비교하여 더 수익이 높은지 판단하는 내용입니다. 2년 뒤의 데이터를 가지고 있을 때만 나올 수 있는 특성을 사용하여 2년 뒤의 내용이 없어도 판단할 수 있는 모델을 만들고자 하였습니다.

**베이스라인 모델과 평가지표를 선택한 이유** : 베이스라인은 1, 즉 투자하는 것으로 선택했습니다. 타겟 특성의 최빈값이 1이기 때문입니다. 또한 평가지표는 recall값을 활용했습니다. 투자한 매물에 대한 투자가치가 중요하기 때문입니다.

**만들고자 하는 모델의 한계, 데이터 누수, 장점, 목표** : 모델은 2년동안의 부동산 매물의 가격과 금리, 등의 데이터를 가지고 만들지만 실제로 2년 뒤의 상황을 알 수 없다는 한계가 있습니다. 이 모델이 가지고 있는 장점은 투자자의 시간을 단축시켜준다는 것에 있습니다. 보다 좋은 매물이 있는 지역과 시기를 추측하여 인력이 좀 더 효율적으로 쓰일 수 있도록 돕는 역할을 합니다. 이 모델의 목표는 현재 가지고 있는 정보를 토대로 2년 뒤 상황에 대해서 추측하는 것입니다.

# CV처리단계

In [12]:
#!pip install category_encoders

In [13]:
from category_encoders import OneHotEncoder
from sklearn.model_selection import cross_val_score
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBClassifier

from scipy.stats import randint, uniform
from sklearn.model_selection import RandomizedSearchCV

from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import cross_val_score

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


## baseline score

In [14]:
print(y_train.value_counts(normalize=True), y_val.value_counts(normalize=True), sep="\n")

1.0    0.585938
0.0    0.414062
Name: 판단여부, dtype: float64
1.0    0.804688
0.0    0.195312
Name: 판단여부, dtype: float64


In [15]:
y_pred_train_base = [1]*len(y_train)
y_pred_val_base = [1]*len(y_val)

recall_train_base = 0.585938
precision_train_base = 1
recall_train_base, precision_train_base

recall_val_base = 0.804688
precision_val_base = 1
recall_val_base, precision_val_base

(0.804688, 1)

##xgboosts 모델 score

In [ ]:
?????????????????????????????????????????????

In [17]:
print('최적 하이퍼파라미터: ', model_xgbc.best_params_)
print('MAE: ', -model_xgbc.best_score_)

최적 하이퍼파라미터:  {'n_estimators': 200, 'max_depth': 12, 'learning_rate': 0.05}
MAE:  0.24182684145996416


In [74]:
best_model_xgbc = model_xgbc.best_estimator_
y_pred_xgbc = best_model_xgbc.predict(X_val_encoded)
mae = mean_absolute_error(y_val, y_pred_xgbc)
print(f'테스트 MAE: {mae:,.7f}')

temp1 = pd.DataFrame(y_val).reset_index().drop("index",axis=1)
temp2 = pd.DataFrame(y_pred_xgbc)
temp = pd.concat([temp1,temp2],axis=1)
temp3 = pd.DataFrame(temp.value_counts())
temp3

테스트 MAE: 0.1145833


0
판단여부 0       
1.0  1.0  309
0.0  1.0   44
     0.0   31

In [79]:
tp = temp3.loc[1].loc[1][0]
fp = temp3.loc[1].loc[0][0]
tn = temp3.loc[0].loc[0][0]
try:
  fn = temp3.loc[0].loc[1][0]
except:
  fn = 0
recall_xgb = tp/(fp+tp)
precision_xgb = tp/(fn+tp)
recall_xgb, precision_xgb

(0.8753541076487252, 1.0)

In [20]:
y_pred_xgbc = best_model_xgbc.predict(X_train_encoded)
mae = mean_absolute_error(y_train, y_pred_xgbc)
print(f'테스트 MAE: {mae:,.7f}')

temp1 = pd.DataFrame(y_train).reset_index().drop("index",axis=1)
temp2 = pd.DataFrame(y_pred_xgbc)
temp = pd.concat([temp1,temp2],axis=1)
pd.DataFrame(temp.value_counts())

테스트 MAE: 0.0000000


0
판단여부 0        
1.0  1.0  1350
0.0  0.0   954

In [25]:
scores_xgb = cross_val_score(best_model_xgbc, X_train_encoded, y_train, cv=10, scoring='recall')
scores_xgb.mean(), scores_xgb.std()

(0.8422222222222222, 0.283173278442975)

##randomforesst 모델 score

In [22]:
pipe = make_pipeline(
    SimpleImputer(), 
    StandardScaler(),
    RandomForestClassifier(random_state=42)
)

dists = {  
    'simpleimputer__strategy': ['mean', 'median'], 
    'randomforestclassifier__n_estimators': randint(50, 300), 
    'randomforestclassifier__max_depth': [5, 10, 15, 20, 50], 
    'randomforestclassifier__max_features': uniform(0, 1)
}

model_rf = RandomizedSearchCV(
    pipe, 
    param_distributions=dists, 
    n_iter=50, 
    cv=3, 
    scoring='neg_mean_absolute_error',  
    verbose=1,
    n_jobs=-1
)

model_rf.fit(X_train_encoded, y_train);

Fitting 3 folds for each of 50 candidates, totalling 150 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  46 tasks      | elapsed:   16.0s
[Parallel(n_jobs=-1)]: Done 150 out of 150 | elapsed:   48.4s finished


In [23]:
print('최적 하이퍼파라미터: ', model_rf.best_params_)
print('MAE: ', -model_rf.best_score_)

최적 하이퍼파라미터:  {'randomforestclassifier__max_depth': 10, 'randomforestclassifier__max_features': 0.0836486992812081, 'randomforestclassifier__n_estimators': 170, 'simpleimputer__strategy': 'mean'}
MAE:  0.3025173611111111


In [81]:
best_model_rf = model_rf.best_estimator_
y_pred_rf = best_model_rf.predict(X_val_encoded)
mae = mean_absolute_error(y_val, y_pred_rf)
print(f'테스트 MAE: {mae:,.7f}')

temp1 = pd.DataFrame(y_val).reset_index().drop("index",axis=1)
temp2 = pd.DataFrame(y_pred_rf)
temp = pd.concat([temp1,temp2],axis=1)
temp3 = pd.DataFrame(temp.value_counts())
temp3

테스트 MAE: 0.1614583


0
판단여부 0       
1.0  1.0  309
0.0  1.0   62
     0.0   13

In [82]:
tp = temp3.loc[1].loc[1][0]
fp = temp3.loc[1].loc[0][0]
tn = temp3.loc[0].loc[0][0]
try:
  fn = temp3.loc[0].loc[1][0]
except:
  fn = 0
recall_rf = tp/(fp+tp)
precision_rf = tp/(fn+tp)
recall_rf, precision_rf

(0.8328840970350404, 1.0)

In [76]:
y_pred_rf = best_model_rf.predict(X_train_encoded)
mae = mean_absolute_error(y_train, y_pred_rf)
print(f'테스트 MAE: {mae:,.7f}')

temp1 = pd.DataFrame(y_train).reset_index().drop("index",axis=1)
temp2 = pd.DataFrame(y_pred_rf)
temp = pd.concat([temp1,temp2],axis=1)
pd.DataFrame(temp.value_counts())

테스트 MAE: 0.0060764


0
판단여부 0        
1.0  1.0  1350
0.0  0.0   940
     1.0    14

In [77]:
scores_rf = cross_val_score(best_model_rf, X_train_encoded, y_train, cv=10, scoring='recall')
scores_rf.mean(), scores_rf.std()

(0.8777777777777779, 0.28142787013822024)

In [84]:
recall_rf, recall_xgb, scores_rf.mean(), scores_xgb.mean()

(0.8328840970350404,
 0.8753541076487252,
 0.8777777777777779,
 0.8422222222222222)

#발표를 위한 주석
**val_평가지표** : 베이스라인에서 계산한 recall값은 0.804688이고 생성한 모델에서 나온 recall값은 XGBClassifier : 0.8753541076487252이고, RandomForestClassifier : 0.8328840970350404이므로 기존의 베이스라인 모델보다 더 향상된 정확도를 가진다.

**교차 검증** : 교차 검증 결과로 recall값은 RandomForestClassifier : 0.8777777777777779이고, XGBClassifier : 0.8422222222222222이므로 RandomForestClassifier가 XGBClassifier보다 더 높은 정확도를 가진다고 볼 수 있다.(val_평가지표와 반대되는 상황 : RandomForestClassifier가 과적합일 가능성이 크다.)

주어진 모델을 가지고 2019~2020년의 투자가치 있는 부동산 추정

In [85]:
y_pred_rf_df = pd.DataFrame(y_pred_rf, columns=["판단여부"])
check_rf = pd.concat([test[["년월","지역","종류"]].reset_index().drop(["index"],axis=1),y_pred_rf_df],axis=1)
check_rf[check_rf["판단여부"]==0.0]

년월    지역     종류  판단여부
21  2019-01-01  강남지역  연립다세대   0.0
22  2019-01-01   서남권  연립다세대   0.0
53  2019-02-01  강남지역  연립다세대   0.0
54  2019-02-01   서남권  연립다세대   0.0
85  2019-03-01  강남지역  연립다세대   0.0
86  2019-03-01   서남권  연립다세대   0.0
117 2019-04-01  강남지역  연립다세대   0.0
118 2019-04-01   서남권  연립다세대   0.0
149 2019-05-01  강남지역  연립다세대   0.0
150 2019-05-01   서남권  연립다세대   0.0
181 2019-06-01  강남지역  연립다세대   0.0
182 2019-06-01   서남권  연립다세대   0.0
245 2019-08-01  강남지역  연립다세대   0.0

추정결과 연립다세대주택을 제외한 모든 부동산은 이득이다.

모델을 학습한 후에 베이스라인보다 잘 나왔나요? 그렇지 않다면 그 이유는 무엇일까요?()
모델 성능 개선을 위해 어떤 방법을 적용했나요? 그 방법을 선택한 이유는 무엇인가요?()
최종 모델에 관해 설명하세요.()

# 추가로 투자하는 부동산 중 어느 지역, 어떤 매물이 더 유리한지 조사해 보았다.

In [86]:
con1 = (sub_df['년월'] < '2018-01-01')
con2 = (sub_df['년월'] >= '2019-01-01')
con3 = ((sub_df['년월'] < '2019-01-01') & (sub_df['년월'] >= '2018-01-01'))

train = sub_df.loc[con1]
test = sub_df.loc[con2]
val = sub_df.loc[con3]

target = '손익분기점'
features = train.columns.drop([target])

X_train = train[features]
y_train = train[target]
X_val = val[features]
y_val = val[target]
X_test = test[features]

X_train_encoded = pd.get_dummies(X_train).drop("년월",axis=1)
X_test_encoded = pd.get_dummies(X_test).drop("년월",axis=1)
X_val_encoded = pd.get_dummies(X_val).drop("년월",axis=1)

In [87]:
pipe = make_pipeline(
    SimpleImputer(), 
    StandardScaler(),
    RandomForestRegressor(random_state=42)
)

dists = {  
    'simpleimputer__strategy': ['mean', 'median'], 
    'randomforestregressor__n_estimators': randint(50, 300), 
    'randomforestregressor__max_depth': [5, 10, 15, 20, 50], 
    'randomforestregressor__max_features': uniform(0, 1)
}

model = RandomizedSearchCV(
    pipe, 
    param_distributions=dists, 
    n_iter=50, 
    cv=3, 
    scoring='neg_mean_absolute_error',  
    verbose=1,
    n_jobs=-1
)

model.fit(X_train_encoded, y_train);

Fitting 3 folds for each of 50 candidates, totalling 150 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  46 tasks      | elapsed:   23.9s
[Parallel(n_jobs=-1)]: Done 150 out of 150 | elapsed:  1.2min finished


In [ ]:
print('최적 하이퍼파라미터: ', model.best_params_)
print('MAE: ', -model.best_score_)

In [ ]:
best_model = model.best_estimator_
y_pred = best_model.predict(X_val_encoded)
mae = mean_absolute_error(y_val, y_pred)
print(f'테스트 MAE: {mae:,.7f}')

In [ ]:
temp1 = pd.DataFrame(y_val).reset_index().drop("index",axis=1)
temp2 = pd.DataFrame(y_pred)
temp = pd.concat([temp1,temp2],axis=1)
temp

In [ ]:
y_pred = best_model.predict(X_test_encoded)
y_pred = pd.DataFrame(y_pred, columns=["손익분기점"])
index_test = pd.concat([test[["년월","지역","종류"]].reset_index().drop(["index"],axis=1),y_pred],axis=1)
index_test[index_test["손익분기점"]>=0.07]

In [ ]:
index_test.sort_values("손익분기점").tail(50)

투자가치가 높은 매물은 아파트로 결론이 나왔으며 지역으로는 동남권, 강남지역, 서남권 순으로 나왔다.

In [ ]:
index_test.sort_values("손익분기점").head(50)

앞에서 나왔던 결론과 마찬가지로 연립다세대 주택은 모든 지역에서 투자가치가 낮은 매물이라는 결론이 나왔다.